src/core/utilities/generate_aliases.ts

```diff
@@ -1,3 +1,133 @@
+function cartesianProduct<T>(arr: T[][]): T[][] {
+  return arr.reduce(
+    (a, b) => {
+      return a
+        .map(x => {
+          return b.map(y => {
+            return x.concat(y);
+          });
+        })
+        .reduce((c, d) => c.concat(d), []);
+    },
+    [[]] as T[][]
+  );
+}
+
+export function* aliasesFromPattern(query: string) {
+  if (!query || query.length === 0) {
+    throw new Error('Tried to parse null or empty string');
+  }
+
+  let i = 0;
+  const currentChar = () => query[i];
+
+  function match(c: string) {
+    const char = currentChar();
+    if (c.includes(char)) {
+      i++;
+      return char;
+    }
+    else if (i >= query.length) {
+      throw new Error(
```

Comment:
Consider throwing your own class here that inherits from Error. Then when you catch, use instanceof operator to detect your error. Re-throw other errors.